### PR TITLE
Rn/test case logcats

### PR DESCRIPTION
**Pull request by rnakade**
Adding new endpoint to get execution steps for individual testcases:
@GET("projects/{projectId}/histories/{historyId}/executions/{executionId}/steps")

Currently in ftl, test case logcats have names like 0000_logcat, 0001_logcat etc.
With this endpoint, we can get the information required to identify the testcases that represent the log files. This info is helpful while uploading test artifacts to ants

### CODE REVIEW

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+internal class TestExecutionStore(
+    private val toolsResultApi: ToolsResultApi,
+    private val googleCloudApi: GoogleCloudApi,
+    firebaseProjectId: String,
+    datastoreApi: DatastoreApi,
+    firebaseTestLabApi: FirebaseTestLabApi,
+    gcsResultPath: String
+) {
+    private val testMatrixStore = TestMatrixStore(
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:16]: why don't we inject these in the constructor.
`TestRunnerServiceImpl` already have them so it should pass them in while creating the execution store.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        firebaseProjectId = firebaseProjectId,
+        datastoreApi = datastoreApi,
+        firebaseTestLabApi = firebaseTestLabApi,
+        toolsResultApi = toolsResultApi,
+        resultsGcsPrefix = googleCloudApi.getGcsPath("aosp-ftl/$gcsResultPath")
+    )
+    private val testLabController = FirebaseTestLabController(
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:23]: ditto

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        testMatrixStore = testMatrixStore
+    )
+
+    private suspend fun getTestExecutionStepsLogcats(
+        projectId: String,
+        historyId: String,
+        executionId: String
+    ): Map<String, String> {
+        val steps = toolsResultApi.listSteps(projectId, historyId, executionId).steps
+        val output = HashMap<String, String>()
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:35]: instead of filling into a map, you can use functional operators

something like:

```
return steps?.flatMap {
    step.testExecutionStep?.toolExecution?.toolOutputs ?: emptylist()
}.mapNotNull {
    if (it.testCase != null) { ... } else null
}
```

lmk if you need more help wrt kotlin parts.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestResultDownloader.kt [hunk 4]
```diff
@@ -66,7 +67,7 @@ internal class TestResultDownloader(
                             target = downloadFolder,
                             filter = { name ->
                                 // these are logs per test, they are plenty in numbers so lets not download them
-                                !name.contains("test_cases")
+                                includeTestCaseArtifacts || !name.contains("test_cases")
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestResultDownloader.kt:70]: this will increase the numbers quite a lot. (log file per test, we have 50-60K of them)

i need to see the androidx.dev side but i was thinking that we can make androidx.dev pass everything we need (FTL Test matrix) and we can return it the log file without re-discovering all of it in TestExecutionStore.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            step.testExecutionStep?.toolExecution?.toolOutputs?.forEach {
+                if (it.testCase != null) {
+                    it.output?.fileUri?.let { it1 -> output.put(it1, it.testCase.className + '_' + it.testCase.name) }
+                }
+            }
+        }
+        return output
+    }
+
+    suspend fun getTestExecutionStepsLogcats(
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:46]: we probably want to return a structured data here (instead of a hashmap, actual class).
also, we likely want to ask them to pass in FTL TestMatrix instance such that we can grab the projectId/historyId from there and query using it.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        testMatrixId: String
+    ): Map<String, String> {
+        val testMatrix = testLabController.getTestMatrix(testMatrixId)
+        testMatrix?.projectId?.let {
+            testMatrix.resultStorage.toolResultsExecution?.historyId?.let { it1 ->
+                testMatrix.resultStorage.toolResultsExecution?.executionId?.let { it2 ->
+                    return getTestExecutionStepsLogcats(it, it1, it2)
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:53]: it is more readable to name these instead of it1/2/3

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerService.kt [hunk 7]
```diff
@@ -91,6 +92,11 @@ interface TestRunnerService {
      */
     suspend fun getTestMatrixResults(testMatrix: TestMatrix): List<TestRunResult>?
 
+    /**
+     * Downloads the file for the given [inputFile] GcsPath to the outputFile.
+     */
+    suspend fun downLoadFileFromGcloud(outputFile: File, inputFile: GcsPath)
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerService.kt:98]: we should return a stream instead of asking an output file so that we can use it w/o saving it to somewhere on the androidx.dev side.

you can check how we read the test result xml files.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerService.kt [hunk 8]
```diff
@@ -275,5 +281,10 @@ interface TestRunnerService {
          * XML result files produced by the test.
          */
         val xmlResults: List<ResultFileResource>
+
+        /**
+         * Test case log files produced by the test.
+         */
+        val testCaseLogcats: Map<String, ResultFileResource>
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerService.kt:288]: as mentioned above, we likely want to avoid returning this here (as it will be big).
instead, query test by test when we are reporting it. Though i'm not sure how we'll associate test with log on the androidx.dev / ANts side. (so my comment my change based on that end)

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt [hunk 9]
```diff
@@ -55,6 +57,17 @@ internal class TestRunnerServiceImpl internal constructor(
...
     )
+    private val testExecutionStore = TestExecutionStore(
+        toolsResultApi = toolsResultApi,
+        googleCloudApi = googleCloudApi,
+        firebaseTestLabApi = firebaseTestLabApi,
+        firebaseProjectId = firebaseProjectId,
+        datastoreApi = datastoreApi,
+        gcsResultPath = gcsResultPath
+    )
+    private val testResultDownloader = TestResultDownloader(
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:68]: for androidx.dev, we shouldn't need this ideally we only use streams there (downloading files is needed for the github Action)

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt [hunk 10]
```diff
@@ -123,7 +136,8 @@ internal class TestRunnerServiceImpl internal constructor(
     }
 
     internal suspend fun findResultFiles(
-        resultPath: GcsPath
+        resultPath: GcsPath,
+        testMatrixId: String? = null
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:140]: see comments above, i don't think we want to be returning all log file references here.

**Review by yigit — CHANGES_REQUESTED** [review]

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestResultDownloader.kt (hunk 4 above)
Comment (rnakade) [reply AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestResultDownloader.kt:70]: Right, I am not really downloading all logs at once on androidx.dev side. I just added this option for later if we want to do bulk upload of artifacts. (As of now, I have no idea how to do bulk upload for artifacts).

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt (hunk 3 above)
Comment (rnakade) [reply AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt]: Changed it to return "step" here

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt [hunk 11]
```diff
@@ -0,0 +1,30 @@
+package dev.androidx.ci.testRunner
```
Comment (liutikas) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:1]: File is missing a copyright header

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt [hunk 12]
```diff
@@ -218,25 +251,30 @@ internal class TestRunnerServiceImpl internal constructor(
...
+        }
 
         override fun toString(): String {
             return """
                 TestResultFiles(
                   device='$deviceRun',
                   logcat=$logcat,
                   intrumentationResult=$instrumentationResult,
-                  xmlResults=$xmlResults,
+                  xmlResults=$xmlResults
```
Comment (liutikas) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:277]: Unrelated change, undo

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt [hunk 13]
```diff
@@ -162,6 +172,27 @@ internal class TestRunnerServiceImpl internal constructor(
...
             } else if (fileName == INSTRUMENTATION_RESULTS_FILE_NAME) {
                 getTestResultFiles(visitor).instrumentationResult = ResultFileResourceImpl(visitor)
+            } else if (fileName.endsWith(LOGCAT_FILE_NAME_SUFFIX)) {
+                val step = steps?.flatMap {
+                    it.testExecutionStep?.toolExecution?.toolOutputs!!
+                }?.find { tor ->
+                    (tor.output?.fileUri == visitor.gcsPath.toString())
+                }
+                val fileUri = step?.output?.fileUri
+                var attemptnumber = 0
```
Comment (liutikas) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:182]: change to camel-case - attemptNumber

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+        projectId: String,
+        historyId: String,
+        executionId: String
+    ): List<Step>? {
+        return toolsResultApi.listSteps(projectId, historyId, executionId).steps
+    }
+
+    suspend fun getTestExecutionSteps(
+        testMatrix: TestMatrix
+    ): List<Step>? {
```
Comment (liutikas) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:20]: Why is this a nullable list? Should we always return an empty list?

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestResultDownloader.kt (hunk 4 above)
Comment (liutikas) [reply AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestResultDownloader.kt:70]: we should only upload logcat for tests that have failed. There is no value in logcat for passing methods.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerService.kt [hunk 15]
```diff
@@ -223,6 +223,20 @@ interface TestRunnerService {
             )
         }
     }
+    data class TestIdentifier(
+        /**
+         * The name of the class.
+         */
+        public val className: String? = null,
```
Comment (liutikas) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerService.kt:230]: Why are these nullable? It seems like we create these outselves and having TestIdentifier that has a null value for either className or name would mean we cannot use this identifier to determine what it points to. Lets make them nonnull.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt [hunk 16]
```diff
@@ -146,6 +154,8 @@ internal class TestRunnerServiceImpl internal constructor(
         // redfin-30-en-portrait_rerun_1/test_result_1.xml
         // redfin-30-en-portrait_rerun_2/logcat
         // redfin-30-en-portrait_rerun_2/test_result_1.xml
+        val testMatrix = testMatrixId?.let { testLabController.getTestMatrix(it) }
```
Comment (liutikas) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:157]: Is it a valid state for us to get  to this and still have null values here?

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt (hunk 11 above)
Comment (rnakade) [reply AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:1]: @liutikas Is this generated from any tool, or can I just copy it from a different file?

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/gcloud/GoogleCloudApi.kt [hunk 17]
```diff
@@ -224,7 +224,7 @@ private class GoogleCloudApiImpl(
 /**
  * Provides access to a Blob returned from a [GoogleCloudApi.walkEntires] method.
  */
-internal interface BlobVisitor {
+interface BlobVisitor {
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/gcloud/GoogleCloudApi.kt:227]: why did we need to open this up?

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt (hunk 11 above)
Comment (yigit) [reply AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt:17]: for the future, you should be able to run ktFormat to fix them (and other files) automatically.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt [hunk 18]
```diff
@@ -162,6 +172,34 @@ internal class TestRunnerServiceImpl internal constructor(
                 getTestResultFiles(visitor).logcat = ResultFileResourceImpl(visitor)
             } else if (fileName == INSTRUMENTATION_RESULTS_FILE_NAME) {
                 getTestResultFiles(visitor).instrumentationResult = ResultFileResourceImpl(visitor)
+            } else if (fileName.endsWith(LOGCAT_FILE_NAME_SUFFIX)) {
+                val step = steps.flatMap {
+                    it.testExecutionStep?.toolExecution?.toolOutputs!!
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:177]: is this guaranteed to be not null?

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt [hunk 19]
```diff
@@ -162,6 +172,34 @@ internal class TestRunnerServiceImpl internal constructor(
...
+                }?.find {
+                    (it.output?.fileUri == visitor.gcsPath.toString())
+                }
+                val fileUri = step?.output?.fileUri
+                var attemptNumber = 0
+                if (fileUri?.contains("rerun_1") == true) {
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:183]: you need to parse these properly, there might be N number of reruns. see the code which parses the outputs where we already have the regex's to do that properly.

**Review by yigit — CHANGES_REQUESTED** [review]

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt (hunk 18 above)
Comment (rnakade) [reply AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:177]: I guess not. I'll update this.

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerService.kt (hunk 8 above)
Comment (rnakade) [reply AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerService.kt]: Leaving it as a map for now. I'll update this when we change the implementation to use the other FTL api: https://firebase.google.com/docs/test-lab/reference/toolresults/rest/v1beta3/projects.histories.executions.steps.testCases/list

File: AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt [hunk 20]
```diff
@@ -188,18 +220,20 @@ internal class TestRunnerServiceImpl internal constructor(
...
         private const val MERGED_TEST_RESULT_SUFFIX = "-test_results_merged.xml"
         private const val LOGCAT_FILE_NAME = "logcat"
+        private const val LOGCAT_FILE_NAME_SUFFIX = "_logcat"
         private const val TEST_RESULT_XML_PREFIX = "test_result_"
         private const val TEST_RESULT_XML_SUFFIX = ".xml"
         private const val INSTRUMENTATION_RESULTS_FILE_NAME = "instrumentation.results"
     }
 
-    private class ResultFileResourceImpl(
+    @JsonClass(generateAdapter = true)
```
Comment (yigit) [inline AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt:235]: i don't think we use generated adapters in this project, it is all reflection so you can remove this.

File: AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/fake/FakeToolsResultApi.kt [hunk 21]
```diff
@@ -48,4 +49,16 @@ internal class FakeToolsResultApi : ToolsResultApi {
         )
         return created
     }
+
+    override suspend fun listSteps(
```
Comment (yigit) [inline AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/fake/FakeToolsResultApi.kt:53]: would be good to add a test for this.

**Review by yigit — COMMENTED** [review]
overall lgtm, just minor nits.

File: AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/FirebaseTestLabControllerTest.kt [hunk 22]
```diff
@@ -99,6 +100,16 @@ internal class FirebaseTestLabControllerTest {
         }
     }
 
+    @Test
```
Comment (yigit) [inline AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/FirebaseTestLabControllerTest.kt:103]: i don't think this test is adding any value. it just makes sure API is there and getting what the fake API will return.

rather, we need a proper test for `AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestExecutionStore.kt` that fills in the fake service (like we do for histories) and makes sure it does pull them properly (e.g. inclluding paging them all).

Comment (rnakade) [reply AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/FirebaseTestLabControllerTest.kt:103]: Can you point to the tests for histories? I see some tests in "ToolsResultApiPlaygroundTest", but those are not based on fake results if I understand correctly.

Comment (rnakade) [reply AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/FirebaseTestLabControllerTest.kt]: @yigit I have added a method to mock some data around steps and and a test to verify it. And also another test in `AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/firebase/ToolsResultApiPlaygroundTest.kt`, I wasn't able to run it through the playground, only verified it though FTL apis directly, but pretty sure it'll work correctly. If you can point me to the "histories" test, I can add something similar for test execution steps too.

Comment (yigit) [reply AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/FirebaseTestLabControllerTest.kt]: Playground tests are not real tests. They are just convenient ways to call the real API to see the response types etc ( to be used during development or troubleshooting). It is also why they are ignored on CI.

As for the actual testing, there are various Fake implementations you can check. The history example was from `FakeToolsResultApi`. Or you can check tests using `FakeFirebaseTestLabApi`. 

For example, `FakeFirebaseTestLabApi` provides `setTestMatrix` which allows the test the update the "backend" TestMatrix value then invoke other methods that should ideally use that new value.

So in this case, i was thinking you expand the FakeToolsResultApi to provide some API to add steps and also be able to return them. Then you can do an end to end test which will need to query these steps and return them (to validate the result)

Another example is `TestRunnerTest`. For instance, `singleTest` creates an archive file on the fake backend, then invokes `testRunner.runTests()` which should be getting those fake zip file and creating tests out of it. Then the test validates the results from the fake backend.

So tl;dr; what I'm looking for is a test that prepares the backend state (add steps etc), run the top level API we have that needs to access to those steps and ensure that returned values include those steps. Which will also allow you to test cases where a result doesn't exist etc.

Comment (rnakade) [reply AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/FirebaseTestLabControllerTest.kt]: @yigit Thank you for these references! I have updated `FakeToolsResultApi` to have a way for adding executionSteps for our tests. I also updated tests in `TestRunnerServiceImplTest` that run more end to end scenarios, with mock data that is inline with what FTL provides through its apis. It covers a positive case where testSteps are present in the api output and logcats are present in the glcoud bucket, as well as negative cases where either step is missing or actual logcat file is missing from gcloud bucket. I have also updated the test with shards and reruns and made sure that scenario is also working as expected.

File: AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/fake/FakeToolsResultApi.kt [hunk 23]
```diff
@@ -59,9 +60,35 @@ internal class FakeToolsResultApi : ToolsResultApi {
...
+            steps = steps[
+                ExecutionStepIdentifier(
+                    projectId = projectId,
+                    historyId = historyId,
+                    executionId = executionId
+                )
+            ]
         )
     }
+    fun createSteps(projectId: String, historyId: String, executionId: String, step: Step?) {
```
Comment (yigit) [inline AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/fake/FakeToolsResultApi.kt:72]: maybe call this `addStep` and set a default value for the step parameter:

```
fun createSteps(projectId: String, historyId: String, executionId: String, step: Step = Step(
   stepId = UUID.randomUUID().toString()
))
```

File: AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/fake/FakeToolsResultApi.kt [hunk 24]
```diff
@@ -59,9 +60,35 @@ internal class FakeToolsResultApi : ToolsResultApi {
...
+        )
+        steps.putIfAbsent(
+            executionStepIdentifier,
+            mutableListOf()
+        )
+        if (step != null) {
+            steps[executionStepIdentifier]?.add(step)
+        } else {
+            steps[executionStepIdentifier]?.add(Step(stepId = UUID.randomUUID().toString()))
+        }
```
Comment (yigit) [inline AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/fake/FakeToolsResultApi.kt:86]: you can rewrite this as:

```
steps.getOrPut(executionStepIdentifier) {
    mutableListOf()
}.add(step)
```
(after also making step non-null.

File: AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/fake/FakeToolsResultApi.kt [hunk 25]
```diff
@@ -59,9 +60,35 @@ internal class FakeToolsResultApi : ToolsResultApi {
...
+    }
+
+    internal data class ExecutionStepIdentifier(
```
Comment (yigit) [inline AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/fake/FakeToolsResultApi.kt:89]: internal or private?

File: AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/TestExecutionStoreTest.kt [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+            Step(stepId = UUID.randomUUID().toString()),
+            Step(stepId = UUID.randomUUID().toString())
+        )
+
+        val projectId = testMatrix.projectId!!
+        val historyId = testMatrix.resultStorage.toolResultsExecution?.historyId!!
+        val executionId = testMatrix.resultStorage.toolResultsExecution?.executionId!!
+
+        // Add 2 steps for this execution
+        inputSteps.forEach() {
```
Comment (yigit) [inline AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/TestExecutionStoreTest.kt:87]: nit, remove `()` from `forEach`)

File: AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/TestExecutionStoreTest.kt [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+                projectId = projectId,
+                historyId = historyId,
+                executionId = executionId,
+                step = it,
+            )
+        }
+        val outputSteps = testExecutionStore.getTestExecutionSteps(testMatrix)
+        assertThat(outputSteps).hasSize(2)
+        assertThat(outputSteps.first().stepId).isEqualTo(inputSteps.first().stepId)
+        assertThat(outputSteps.last().stepId).isEqualTo(inputSteps.last().stepId)
```
Comment (yigit) [inline AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/TestExecutionStoreTest.kt:98]: does this not work?

```
assertThat(outputSteps).containsExactlyElementsIn(inputSteps)
```

File: AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImplTest.kt [hunk 28]
```diff
@@ -323,6 +371,56 @@ class TestRunnerServiceImplTest {
                 "test_result_1 content xml",
                 "test_result_2 content xml",
             )
+
+            assertThat(
+                testRun.testCaseLogcats.size
+            ).isEqualTo(
+                1
+            )
+            // step and logcat both have valid values for test1
```
Comment (yigit) [inline AndroidXCI/lib/src/test/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImplTest.kt:380]: yep, this is exactly what i was looking for 👍

**Review by yigit — CHANGES_REQUESTED** [review]
lgtm, just minor nits.

**Comment by rnakade** [comment]
@yigit Addressed all the feedack. Yay!

**Review by yigit — APPROVED** [review]
